### DISCUSSION TITLE
[Proposal] Autofetching method arguments via object deconstruction

### DISCUSSION THREAD

**Discussion by Horusiath** [dotnet/csharplang#1648] [Language Ideas]
The idea presented here aims to introduce a simpler mechanics of supplying arguments of methods and constructors using object deconstruction pattern. It could simplify things like non-destructive mutations on record types, easier mapping from one type to another.

The syntax example (using C# records proposal).

```csharp
// C# record class - it implies that this class defines method
// `void Deconstruct(out int id, out string firstName, out string lastName)`
class User(int Id, string FirstName, string LastName);  
var user = new User(1, "John", "Doe");

string FullName(string firstName, string lastName) => firstName + " " + lastName;

// new syntax proposal
string johnDoeFullName = FullName(...user);  // "John Doe"
// fetch with override
string johnSmithFullName = FullName(...user, lastName: "Smith");   // "John Smith"
```

### How could it work

The "fetching" operator is actually `...` mark - what it does is to execute `Deconstruct` upon object or structure it targets, and then fetches assigned `out` parameters from it with parameters from called method or constructor:

1. Fetching is based on parameter name and type.
2. All that fetching feature requires, is a `Deconstruct` method to be defined:
    - This way we can create custom Deconstruct methods that will be able to be fetched into the methods we desire even if they originally weren't compatible for it.
    - It also allows us to prevent from accidental parameter fetching over the fields/properties, we don't want to use with that feature.
3. Fetching is type safe - in case if not all of the called method arguments were supplied by a given `Deconstruct` method, compiler must inform about missing arguments in error message. In that case programmer must explicitly supply missing arguments.
4. Fetching is call-side sensitive - it means that we need to match only the arguments needed by a called method. As in example above: even thou `User.Deconstruct` supplied id parameter as well as first and last name, id is being ignored, as it doesn't match with any of the arguments of FullName method call.

### Use case 1: `With` method

One of the examples of such use case would be a [With](https://github.com/dotnet/csharplang/blob/master/proposals/records.md#with-expressions) expression proposal. As its behavior is well defined in records representing sealed types, existence of `With` methods is more problematic in situations where records themselves can be used with inheritance.

This feature would allow to simplify that process. Taken example from F#:

```fsharp
type User = { Id: int; FirstName: string; LastName: string }
let alice = { Id = 1; FirstName = "Alice"; LastName = "Stone" }
let jane = { alice with Id = 2; FirstName = "Jane" }
```

this could be roughly represented as equivalent:

```csharp
class User(int Id, string FirstName, string LastName); 
var alice = new User(1, "Alice", "Stone");
var jane = new User(...alice, id: 2, firstName: "Stone");
```

but what's even more important, we could use it over type inheritance:

```csharp
class Employee(int Id, string FirstName, string LastName, string Company) : User(Id, FirstName, LastName);
var alice = new User(1, "Alice", "Stone");
// as company is missing it must be supplied manually
var aliceAsTeacher = new Employee(...alice, "University of Toronto");
// and back to User type
alice = new User(...aliceAsTeacher);
```

### Use case 2: mapping between types

Many developers often find themselves in a position where they need to map between two very similar data types from different domains. This is so common case, that specific libraries where developed to handle this case (i.e: AutoMapper). This syntax would allow to map simple cases automatically and let them to be compile time checked.

Example:

```csharp
class User(int Id, string FirstName, string LastName, Address address);
class UserInfo(string FirstName, string LastName);

var alice = new User(1, "Alice", "Stone", new Address("Some street 2", "Toronto", "Canada"));
var aliceDto = new UserInfo(...alice);
````

### Things to consider:

- [ ] How should it work with sub-typing: I proposed for argument fetching to match over argument name and type, but how to treat sub- or super-types of a given type?
- [ ] How should it work with types that are subject to implicit cast operators?
- [ ] Since C# allows to specify method overloads, how automatic argument fetching should decide which overload to take? Maybe if 1-1 matching is not possible, it should simply throw compiler error about inconclusive matching?

**Comment by HaloFour** [comment]
Sounds very similar to "splatting" although using deconstruction instead of just tuples.

> Fetching is call-side sensitive - it means that we need to match only the arguments needed by a called method.

This sounds dangerous.

> Fetching is based on parameter name and type.

Currently deconstruction and tuple assignment is entirely positional.  While convenient it would be inconsistent.
